apps/agent/contracts/SignatureValidator.sol

```diff
@@ -0,0 +1,156 @@
+    bytes4 public constant ERC1271_RETURN_VALID_SIGNATURE = 0x20c13b0b;
+    uint256 internal constant ERC1271_ISVALIDSIG_MAX_GAS = 250000;
+
+    string private constant ERROR_INVALID_LENGTH_POP_BYTE = "SIGVAL_INVALID_LENGTH_POP_BYTE";
+
+    /// @dev Validates that a hash was signed by a specified signer.
+    /// @param hash Hash which was signed.
+    /// @param signer Address of the signer.
+    /// @param signature ECDSA signature along with the mode (0 = Invalid, 1 = EIP712, 2 = EthSign, 3 = ERC1271) {mode}{r}{s}{v}.
+    /// @return Returns whether signature is from a specified user.
+    function isValidSignature(bytes32 hash, address signer, bytes signature) internal view returns (bool) {
+        if (signature.length == 0) {
+            return false;
+        }
+
+        uint8 modeByte = uint8(signature[0]);
+        if (modeByte >= uint8(SignatureMode.NMode)) {
+            return false;
+        }
+        SignatureMode mode = SignatureMode(modeByte);
+
+        if (mode == SignatureMode.EIP712) {
+            return ecVerify(hash, signer, signature);
+        } else if (mode == SignatureMode.EthSign) {
+            return ecVerify(
+                keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash)),
+                signer,
+                signature
+            );
+        } else if (mode == SignatureMode.ERC1271) {
+            // Pop the mode byte before sending it down the validation chain
+            return safeIsValidSignature(signer, hash, popFirstByte(signature));
+        } else {
+            return false;
+        }
+    }
+
+    function ecVerify(bytes32 hash, address signer, bytes memory signature) private pure returns (bool) {
+        (bool badSig, bytes32 r, bytes32 s, uint8 v) = unpackEcSig(signature);
+
+        if (badSig) {
+            return false;
+        }
+
+        return signer == ecrecover(hash, v, r, s);
+    }
+
+    function unpackEcSig(bytes memory signature) private pure returns (bool badSig, bytes32 r, bytes32 s, uint8 v) {
+        if (signature.length != 66) {
+            badSig = true;
+            return;
+        }
+
+        v = uint8(signature[65]);
+        assembly {
+            r := mload(add(signature, 33))
+            s := mload(add(signature, 65))
+        }
+
+        // Allow signature version to be 0 or 1
+        if (v < 27) {
+            v += 27;
+        }
+
+        if (v != 27 && v != 28) {
+            badSig = true;
+        }
+    }
+
+    function popFirstByte(bytes memory input) private pure returns (bytes memory output) {
+        uint256 inputLength = input.length;
+        require(inputLength > 0, ERROR_INVALID_LENGTH_POP_BYTE);
+
+        output = new bytes(inputLength - 1);
+
+        uint256 inputPointer;
+        uint256 outputPointer;
+        assembly {
+            inputPointer := add(input, 0x21)
+            outputPointer := add(output, 0x20)
+        }
+        memcpy(outputPointer, inputPointer, inputLength);
```

Comment:
I think there should be `memcpy(outputPointer, inputPointer, inputLength - 1);` since the length of the new array should be `inputLength - 1`